JS: add query: js/useless-defensive-code

esben-semmle opened the pull request:
This PR introduces `js/useless-defensive-code` as a `recommendation` query. The query is based on syntactic patterns, and type inference for a variable in each pattern.

Defensive code is typically a condition that prevents `null` or `undefined` values from causing a crash later on. Defensive code is useless when the condition evaluates to a single value for all execution paths. In that case, the condition might as well be removed. I welcome a discussion of the patterns.

The end result is that:

1.  Some alerts from `js/trivial-conditional` and `js/comparison-between-incompatible-types` are whitelisted
2.  The above whitelisted alerts are flagged by `js/useless-defensive-code`
3.  Additional, more advanced, unnecessary defensive code variants are also flagged by `js/useless-defensive-code`

The most important consequence of this is that the remaining alerts from `js/trivial-conditional` and `js/comparison-between-incompatible-types` now are much more severe on average. A secondary consequence is that alerts caused by type inference improvements will be easier to judge the value of.

Please note that the query deliberately treats cases like `o === undefined || o.p` and `o === undefined && o.p` identically, for the sake of simplicity. It is the responsibility of another query to flag the definite error in the latter case.

Bullet 1 and 2 above may cause a special kind of wobbliness: if we gain precision because of a simpler program or because of an improved analysis implementation, then additional alerts may be flagged by `js/useless-defensive-code`, while a more complex program may change those alerts to `js/trivial-conditional`. C.f. [another evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/always-return-type-inference_1539684685915#--jstrivial-conditional-for-etheretherpad-lite-1).

An oldish [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/useless-defensive-code_1539553435513) shows that performance is fine and an interesting amount of result changes, with a net additional 69 alerts that the programmer will be presented for by default. A smaller, but more recent, [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/useless-defensive-code_1541029994872) on top of the current master shows that this still scales.

asger-semmle commented:
While we're on the subject, am I the only one who thinks the word "useless" sounds rude? I liked the previously suggested "unnecessary defensive programming".

Do I understand it correctly that this will not introduce any new alerts, but merely reclassify existing ones into this less severe query? It's not quite clear from bullet 3 if this means new alerts.

esben-semmle commented:
The phrasing is inherited from `js/trivial-conditional` aka `UselessConditional.ql` which has `@name Useless conditional`. We should change both or none.

---

Brand new alerts will be added because of some additional constant folding steps in this query. Consider the simple case of:

```javascript
    if (u !== undefined) {
        u.p;
    }
```
This is obviously not flagged by `js/heterogeneous-comparison` because the comparison is homogeneous, and it is not flagged by `js/trivial-conditional` because we do not have a type inference step for the result of a comparison. 
Another example is:

```javascript
    typeof f === "function" && f();
```

Maybe we should let `js/trivial-conditional` know that `null` coerces to `false`, but that is the only low hanging fruit I have seen.

xiemaisi commented:
Ping @Semmle/doc on whether "useless" is considered rude; as a non-native speaker I don't feel qualified to judge.

xiemaisi commented on javascript/ql/src/Expressions/UselessDefensiveProgramming.qhelp:
```diff
@@ -0,0 +1,90 @@
+		<p>
+
+			Examine the surrounding code to determine if the defensive
+			code is worth keeping despite being practically useless.
```
...and what is our suggestion if it does turn out to be worth keeping?

xiemaisi commented on javascript/ql/src/Statements/UselessConditional.ql:
```diff
@@ -95,6 +73,19 @@ predicate isConstantBooleanReturnValue(Expr e) {
+  exists(Expr defensive |
+    defensive = e or
+    // traverse negations
+    defensive.(LogNotExpr).getOperand+() = e
```
Doesn't this skip through arbitrary unary expression operands? Is this what we want?

xiemaisi commented on javascript/ql/src/semmle/javascript/DefensiveProgramming.qll:
```diff
@@ -0,0 +1,412 @@
+/**
+ * A test in a defensive programming pattern.
+ */
+abstract class DefensiveExpression extends DataFlow::ValueNode {
```
I'd prefer having the word "test" or "check" in the class name (it already appears in the qldoc comment).

xiemaisi commented on javascript/ql/src/semmle/javascript/DefensiveProgramming.qll:
```diff
@@ -0,0 +1,412 @@
+    }
+
+    override boolean getTheTestResult() {
+      result = getPolarity() and
```
There is some very subtle logic at work here that I don't fully understand. Could you document it a bit more, please?

xiemaisi commented on javascript/ql/src/semmle/javascript/DefensiveProgramming.qll:
```diff
@@ -0,0 +1,412 @@
+
+    SanityCheckingUndefinedNullGuard() {
+      exists (IfStmt c |
+        this = c.getCondition().flow() and
```
What about conditional expressions?

esben-semmle replied:
That is a deliberate omission, based on observed false positives for `js/heterogeneous-comparison`. Note that this test does not have a specific node that it guards. It there was such a node, then the conditional expression would be used, i.e. we flag:

```
if (e !== undefined) { e(); }
if (e !== undefined) { f(); }
e !== undefined? e(): undefined;
```
but not 
```
e !== undefined? f(): undefined;
```

xiemaisi commented on javascript/ql/src/semmle/javascript/DefensiveProgramming.qll:
```diff
@@ -0,0 +1,412 @@
+  private Expr getAGuardedExpr(Expr guard) {
+    exists(BinaryExpr op |
+      op.getLeftOperand() = guard and
+      (op instanceof LogAndExpr or op instanceof LogOrExpr) and
```
`LogicalBinaryExpr`

xiemaisi commented on javascript/ql/src/semmle/javascript/DefensiveProgramming.qll:
```diff
@@ -0,0 +1,412 @@
+    exists(IfStmt c |
+      c.getCondition() = guard |
+      result = c.getAControlledStmt().getChildExpr(0) or
+      result = c.getAControlledStmt().(BlockStmt).getStmt(0).getChildExpr(0)
```
I don't like the use of `getChildExpr(0)` here; it makes assumptions about the order of evaluation that are hard to keep track of. Do you perhaps want to use `getFirstControlFlowNode`?

esben-semmle replied:
Hmm, it is quite awkward to get the first expression since `getFirstControlFlowNode` for the case we care about (`ExprStmt`) is the node itself. See eb9c57cf. We may want a proper `getFirstExpr` at a later time.

esben-semmle commented:
> Overall, the large number of syntactic checks seems quite cumbersome.

Agreed, but I believe we must use syntactic checks to identify the relevant programming patterns.

This could be simplified a bit if we added constant folding for boolean results to the type inference, but we decided that it may not be worth it for just this query and `js/trivial-conditional`. Such a constant folding would also mean that `js/trivial-conditional` needed a whitelist for the `js/heterogenous-comparison`.

> and I'm not confident I fully understand all of the patterns being checked for. 

I have added source code examples to the docstrings.

esben-semmle commented:
Rebased on current master to make the tests for `js/trivial-conditional` work.

mc-semmle commented:
Asked @semmledocs-ac for help as I'm not a native speaker either. Please let me know when this is ready for a doc review. Thanks.

esben-semmle commented:
@mc-semmle I have addressed the "is useless" comments, and this PR is ready for doc review.

mc-semmle commented on javascript/ql/src/Expressions/UselessDefensiveProgramming.ql:
```diff
@@ -0,0 +1,32 @@
+/**
+ * @name Useless defensive code
```
Are we happy with the use of the word "Useless" in the query name?

mc-semmle commented on javascript/ql/src/Expressions/UselessDefensiveProgramming.ql:
```diff
@@ -0,0 +1,32 @@
+/**
+ * @name Useless defensive code
+ * @description If the situation some defensive code guards against never
```
Sorry but I don't understand what you mean in the description. Would it be possible to rephrase so it's clearer?

mc-semmle commented on javascript/ql/src/Expressions/UselessDefensiveProgramming.qhelp:
```diff
@@ -0,0 +1,91 @@
+
+			However, if the situation that some defensive code guards
+			against never can occur, then the defensive code serves no purpose and
+			it can safely be removed.
```
Remove "it"?
...then the defensive code serves no purpose and can safely be removed.

mc-semmle commented on javascript/ql/src/Expressions/UselessDefensiveProgramming.qhelp:
```diff
@@ -0,0 +1,91 @@
+		<p>
+
+			However, due to coercion rules, <code>v ==
+			undefined</code> holds for both the situation where <code>v</code>
```
Same here - remove "it"?
so the <code>v == null</code> guard serves no purpose, and can be removed:

mc-semmle commented:
esben-semmle - doc review completed. This LGTM, I have just made a few minor comments / suggestions. Feel free to ignore them if you don't agree.

esben-semmle commented:
Thank you @mc-semmle 

The query description is now "Defensive code that guards against a situation that never happens is not needed." And the query name is "Unneeded defensive code"

mc-semmle commented:
Thank you for the updates @esben-semmle - LGTM 👍

xiemaisi commented:
Conflicts.

esben-semmle commented:
Conflicts resolved.
